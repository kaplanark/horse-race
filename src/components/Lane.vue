<script setup>
const props = defineProps({
	horse: {
		type: Object,
		default: {}
	},
	laneNo: {
		type: Number,
		default: 0
	}
});
</script>

<template>
	<div class="lane" :lane-no="props.laneNo">
		<div class="horse" :class="{'runing':props.horse.run}"
			:style="{'--color-horse':props.horse.color,'margin-left':props.horse.travelledDistance+'px'}"></div>
	</div>
</template>

<style lang="scss">
.lane {
	position: relative;
	background-image: url('@assets/images/bg-lane.png');
	background-size: contain;
	background-position: bottom;
	background-repeat: no-repeat;
	max-width: 100vw;
	overflow: hidden;

	.horse {
		height: 100%;
		width: 80px;
		position: relative;
		transition: all 0.15s ease-in-out;

		&::before {
			position: absolute;
			content: 'üêé';
			right: 0;
			top: 50%;
			transform: translateY(-65%);
			font: 72px Muybridge;
			line-height: 1;
			color: var(--color-horse);
			z-index: 1;
		}

		&.runing {
			animation: 1s linear infinite gallop;
		}
	}

	&::before {
		position: absolute;
		content: attr(lane-no);
		top: 50%;
		left: 2px;
		transform: translateY(-50%);
		height: 24px;
		width: 24px;
		color: var(--color-white);
		font-size: 18px;
		display: flex;
		align-items: center;
		justify-content: center;
		z-index: 2;
		background-color: var(--color-primary);
		border-radius: 50%;
	}

	&::after {
		position: absolute;
		content: '';
		top: 50%;
		right: 0;
		height: 70%;
		width: 80px;
		transform: translateY(-50%);
		z-index: 0;
		border-left: 1px dashed var(--color-primary);
		opacity: 0.5;
	}
}
</style>