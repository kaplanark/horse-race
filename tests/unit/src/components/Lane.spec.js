import { describe, expect, test } from 'vitest'
import { mount } from '@vue/test-utils'
import Lane from '@components/Lane.vue';

describe('Lane.vue', () => {
	let wrapper;
	beforeEach(() => {
		wrapper = mount(Lane, {
			props: {
				horse: {
					run: true,
					travelledDistance: 210,
					color: '#3b302f',
				},
				laneNo: 1,
			}
		});
	});

	afterEach(() => {
		wrapper.unmount();
	});

	test('is correctly rendered', () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('should renders is page content is correct', () => {
		expect(wrapper.vm).toBeTruthy();
	});


	test('renders props.horse when passed', () => {
		expect(wrapper.vm.horse).toBeTruthy();
		expect(wrapper.get('.horse').classes()).toContain('running');
		expect(wrapper.get('.horse').attributes('style')).toContain('--color-horse: #3b302f; margin-left: 210px;');
	});

	test('renders props.laneNo when passed', () => {
		expect(wrapper.vm.laneNo).toEqual(1);
		expect(wrapper.attributes('data-lane')).toEqual('1');
	});
});