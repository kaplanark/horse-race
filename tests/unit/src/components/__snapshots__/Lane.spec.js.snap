// Vitest Snapshot v1

exports[`Lane.vue > is correctly rendered 1`] = `
"<div class=\\"lane\\" data-lane=\\"1\\">
  <div class=\\"horse running\\" style=\\"--color-horse: #3b302f; margin-left: 210px;\\"></div>
</div>"
`;
